serenity-bdd update: HttpsProxyAgent causes TypeError

I'm trying to replay the [Serenity/JS Cucumber Template](https://github.com/serenity-js/serenity-js-cucumber-template.git) (at commit https://github.com/serenity-js/serenity-js-cucumber-template/commit/dcd1deae8afefc1c45034a32eefd613ef44a3ece) behind a corporate HTTP(S) Proxy. The version of `serenity-bdd` used therein is **3.6.1**

When doing `npm install` the `postinstall` script runs `serenity-bdd update`, which crashes with the follow trace:

```
ConfigurationError: UpdateCommandActors encountered a problem when preparing actor "Serenity/JS Updater" for stage
    at Stage.actor (/home/dummy/serenity-js-cucumber-template/node_modules/@serenity-js/core/lib/stage/Stage.js:89:23)
    at Stage.theActorCalled (/home/dummy/serenity-js-cucumber-template/node_modules/@serenity-js/core/lib/stage/Stage.js:67:21)
    ... 3 lines matching cause stack trace ...
    at /home/dummy/serenity-js-cucumber-template/node_modules/yargs/build/index.cjs:1:8993
    at j (/home/dummy/serenity-js-cucumber-template/node_modules/yargs/build/index.cjs:1:4956)
    at _.handleValidationAndGetResult (/home/dummy/serenity-js-cucumber-template/node_modules/yargs/build/index.cjs:1:8962)
    at _.applyMiddlewareAndGetResult (/home/dummy/serenity-js-cucumber-template/node_modules/yargs/build/index.cjs:1:9604)
    at _.runCommand (/home/dummy/serenity-js-cucumber-template/node_modules/yargs/build/index.cjs:1:7231)
Caused by: TypeError: HttpsProxyAgent is not a constructor
    at proxiedHttpsAgent (/home/dummy/serenity-js-cucumber-template/node_modules/@serenity-js/serenity-bdd/lib/cli/io/axiosClient.js:122:12)
    at axiosClient (/home/dummy/serenity-js-cucumber-template/node_modules/@serenity-js/serenity-bdd/lib/cli/io/axiosClient.js:54:15)
    at UpdateCommandActors.actors [as createAxios] (/home/dummy/serenity-js-cucumber-template/node_modules/@serenity-js/serenity-bdd/lib/cli/commands/update.js:41:110)
    at UpdateCommandActors.prepare (/home/dummy/serenity-js-cucumber-template/node_modules/@serenity-js/serenity-bdd/lib/cli/stage/UpdateCommandActors.js:16:57)
    at Stage.actor (/home/dummy/serenity-js-cucumber-template/node_modules/@serenity-js/core/lib/stage/Stage.js:84:35)
    at Stage.theActorCalled (/home/dummy/serenity-js-cucumber-template/node_modules/@serenity-js/core/lib/stage/Stage.js:67:21)
    at Serenity.theActorCalled (/home/dummy/serenity-js-cucumber-template/node_modules/@serenity-js/core/lib/Serenity.js:193:27)
    at actorCalled (/home/dummy/serenity-js-cucumber-template/node_modules/@serenity-js/core/lib/instance.js:207:29)
    at Object.handler (/home/dummy/serenity-js-cucumber-template/node_modules/@serenity-js/serenity-bdd/lib/cli/commands/update.js:47:46)
    at /home/dummy/serenity-js-cucumber-template/node_modules/yargs/build/index.cjs:1:8993 {
  cause: TypeError: HttpsProxyAgent is not a constructor
      at proxiedHttpsAgent (/home/dummy/serenity-js-cucumber-template/node_modules/@serenity-js/serenity-bdd/lib/cli/io/axiosClient.js:122:12)
      at axiosClient (/home/dummy/serenity-js-cucumber-template/node_modules/@serenity-js/serenity-bdd/lib/cli/io/axiosClient.js:54:15)
      at UpdateCommandActors.actors [as createAxios] (/home/dummy/serenity-js-cucumber-template/node_modules/@serenity-js/serenity-bdd/lib/cli/commands/update.js:41:110)
      at UpdateCommandActors.prepare (/home/dummy/serenity-js-cucumber-template/node_modules/@serenity-js/serenity-bdd/lib/cli/stage/UpdateCommandActors.js:16:57)
      at Stage.actor (/home/dummy/serenity-js-cucumber-template/node_modules/@serenity-js/core/lib/stage/Stage.js:84:35)
      at Stage.theActorCalled (/home/dummy/serenity-js-cucumber-template/node_modules/@serenity-js/core/lib/stage/Stage.js:67:21)
      at Serenity.theActorCalled (/home/dummy/serenity-js-cucumber-template/node_modules/@serenity-js/core/lib/Serenity.js:193:27)
      at actorCalled (/home/dummy/serenity-js-cucumber-template/node_modules/@serenity-js/core/lib/instance.js:207:29)
      at Object.handler (/home/dummy/serenity-js-cucumber-template/node_modules/@serenity-js/serenity-bdd/lib/cli/commands/update.js:47:46)
      at /home/dummy/serenity-js-cucumber-template/node_modules/yargs/build/index.cjs:1:8993
}
```

If I trace it back I realize that it's `@serenity-js/serenity-bdd/lib/cli/io/axiosClient.js` at line 122 which creates a `HttpsProxyAgent` object and provokes the `TypeError: HttpsProxyAgent is not a constructor`.

As a work-around I can bypass the problematic lines of code by invoking it like this: `serenity-bdd update --repository http://insecure.repo1.maven.org/maven2/`, which overrides `https://repo1.maven.org/maven2/` from `defaults.ts` so that the Axios Client is configured with a HTTP Proxy Agent rather than HTTPS. 

How can this be fixed in a more sustainable and secure way?

--- a/documentation/serenity-js.org/docusaurus.config.js
+++ b/documentation/serenity-js.org/docusaurus.config.js
@@ -101,7 +101,7 @@ const config = {
                 },
                 items: [
                     { label: 'Handbook', type: 'doc', docId: 'index', position: 'left' },
-                    { label: 'API', to: '/api', position: 'left' },
+                    { label: 'API', to: '/api/core', position: 'left' },
                     { to: '/blog', label: 'Blog', position: 'left' },
                     { to: '/contributing', label: 'Contribute', position: 'left' },
                     { label: `Changelog \uD83C\uDF81`, to: 'changelog',  position: 'left' },
@@ -170,7 +170,7 @@ const config = {
                     {
                         title: 'Developers',
                         items: [
-                            { label: 'API Docs', to: '/api' },
+                            { label: 'API Docs', to: '/api/core' },
                             { label: 'Project Templates', to: 'https://github.com/serenity-js/?q=template&type=all&language=&sort=' },
                             { label: 'Serenity/JS GitHub', href: 'https://github.com/serenity-js', },
                             { label: 'Report an issue', href: 'https://github.com/serenity-js/serenity-js/issues', },

--- a/packages/serenity-bdd/spec/cli/io/axiosClient.spec.ts
+++ b/packages/serenity-bdd/spec/cli/io/axiosClient.spec.ts
@@ -163,7 +163,7 @@ describe('axiosClient', () => {
 
         describe('when no proxy is needed', () => {
 
-            it('uses a HTTPS Agent configured to ignored SSL errors', () => {
+            it('uses a HTTPS Agent configured to ignore SSL errors', () => {
                 const env = { npm_config_ca: ca };
                 const ignoreSsl = true;
 
@@ -179,7 +179,7 @@ describe('axiosClient', () => {
                 expect(client.defaults.httpsAgent.options.rejectUnauthorized).to.equal(!ignoreSsl);
             });
 
-            it('uses a HTTPS Agent configured to not ignored SSL errors', () => {
+            it('uses a HTTPS Agent configured to not ignore SSL errors', () => {
                 const env = { npm_config_ca: ca };
                 const ignoreSsl = false;
 

--- a/packages/serenity-bdd/spec/cli/io/axiosClient.spec.ts
+++ b/packages/serenity-bdd/spec/cli/io/axiosClient.spec.ts
@@ -1,0 +1,315 @@
+import { expect } from '@integration/testing-tools';
+import { ConfigurationError } from '@serenity-js/core';
+import { readFileSync } from 'fs';
+import { describe, it } from 'mocha';
+import { resolve } from 'path';
+import { URL } from 'url';
+
+import { axiosClient } from '../../../src/cli/io';
+import { Credentials } from '../../../src/cli/model';
+
+describe('axiosClient', () => {
+
+    const repositoryAuth = new Credentials('username', 'password');
+    const repositoryNoAuth = new Credentials(undefined, undefined);
+
+    it('correctly configures the User-Agent if needed', () => {
+        const repository = new URL('https://artifactory.example.org');
+
+        const ignoreSsl = false;
+        const env = {
+            npm_config_user_agent: 'serenity-js'
+        }
+
+        const client = axiosClient(repository, ignoreSsl, env, repositoryNoAuth);
+
+        expect(client.defaults.baseURL).to.equal(repository.toString());
+        expect(client.defaults.adapter).to.deep.equal([ 'http' ]);
+        expect(client.defaults.auth).to.deep.equal({ username: undefined, password: undefined });
+        expect(client.defaults.proxy).to.equal(false);
+        expect(client.defaults.headers['User-Agent']).to.equal('serenity-js');
+    });
+
+    describe('when no proxy is needed', () => {
+
+        const ignoreSsl = false;
+        const env = {};
+
+        it('correctly configures the defaults', () => {
+            const repository = new URL('https://artifactory.example.org');
+
+            const client = axiosClient(repository, ignoreSsl, env, repositoryNoAuth);
+
+            expect(client.defaults.baseURL).to.equal(repository.toString());
+            expect(client.defaults.adapter).to.deep.equal([ 'http' ]);
+            expect(client.defaults.auth).to.deep.equal({ username: undefined, password: undefined });
+            expect(client.defaults.proxy).to.equal(false);
+        });
+
+        it('correctly configures authentication credentials', () => {
+            const repository = new URL('https://artifactory.example.org');
+
+            const client = axiosClient(repository, ignoreSsl, env, repositoryAuth);
+
+            expect(client.defaults.baseURL).to.equal(repository.toString());
+            expect(client.defaults.adapter).to.deep.equal([ 'http' ]);
+            expect(client.defaults.auth).to.deep.equal({
+                username: repositoryAuth.username,
+                password: repositoryAuth.password,
+            });
+            expect(client.defaults.proxy).to.equal(false);
+        });
+    });
+
+    describe('when the artifact registry works over HTTP', () => {
+
+        const repository = new URL('http://artifactory.example.org');
+        const ignoreSsl = false;
+        const env = {
+            HTTP_PROXY: 'http://proxy.example.org'
+        };
+
+        it('correctly configures a HTTP proxy', () => {
+            const client = axiosClient(repository, ignoreSsl, env, repositoryNoAuth);
+
+            expect(client.defaults.baseURL).to.equal(repository.toString());
+            expect(client.defaults.adapter).to.deep.equal([ 'http' ]);
+            expect(client.defaults.auth).to.deep.equal({ username: undefined, password: undefined });
+
+            expect(client.defaults.proxy).to.deep.equal({
+                protocol: 'http:',
+                host: 'proxy.example.org',
+                port: 80,
+                auth: undefined
+            });
+        });
+
+        it('correctly configures a HTTP proxy with credentials', () => {
+            const client = axiosClient(repository, ignoreSsl, {
+                HTTP_PROXY: 'http://proxy-user:proxy-password@proxy.example.org'
+            }, repositoryNoAuth);
+
+            expect(client.defaults.baseURL).to.equal(repository.toString());
+            expect(client.defaults.adapter).to.deep.equal([ 'http' ]);
+            expect(client.defaults.auth).to.deep.equal({ username: undefined, password: undefined });
+
+            expect(client.defaults.proxy).to.deep.equal({
+                protocol: 'http:',
+                host: 'proxy.example.org',
+                port: 80,
+                auth: {
+                    password: 'proxy-password',
+                    username: 'proxy-user',
+                }
+            });
+        });
+
+        it('complains if the HTTP proxy URL is invalid', () => {
+            expect(() => axiosClient(repository, ignoreSsl, { HTTP_PROXY: 'invalid-url' }, repositoryNoAuth))
+                .to.throw(ConfigurationError, 'Env variable HTTP_PROXY=invalid-url should specify a valid URL');
+        });
+
+        it('complains if the HTTP proxy protocol is not supported', () => {
+            expect(() => axiosClient(repository, ignoreSsl, { HTTP_PROXY: 'invalid://localhost' }, repositoryNoAuth))
+                .to.throw(ConfigurationError, 'Env variable HTTP_PROXY=invalid://localhost should specify protocol to be used, i.e. http:// or https://');
+        });
+
+        it('does not use the proxy if repository is in no_proxy list', () => {
+            const client = axiosClient(repository, ignoreSsl, {
+                ...env,
+                no_proxy: 'artifactory.example.org'
+            }, repositoryNoAuth);
+
+            expect(client.defaults.baseURL).to.equal(repository.toString());
+            expect(client.defaults.adapter).to.deep.equal([ 'http' ]);
+            expect(client.defaults.auth).to.deep.equal({ username: undefined, password: undefined });
+
+            expect(client.defaults.proxy).to.equal(undefined);
+        });
+
+        it('does not use a proxy if repository domain is in no_proxy list', () => {
+            const client = axiosClient(repository, ignoreSsl, {
+                ...env,
+                no_proxy: '.example.org'
+            }, repositoryNoAuth);
+
+            expect(client.defaults.baseURL).to.equal(repository.toString());
+            expect(client.defaults.adapter).to.deep.equal([ 'http' ]);
+            expect(client.defaults.auth).to.deep.equal({ username: undefined, password: undefined });
+
+            expect(client.defaults.proxy).to.equal(undefined);
+        });
+
+        it('does not use a proxy if no_proxy set to *', () => {
+            const client = axiosClient(repository, ignoreSsl, {
+                ...env,
+                no_proxy: '*'
+            }, repositoryNoAuth);
+
+            expect(client.defaults.baseURL).to.equal(repository.toString());
+            expect(client.defaults.adapter).to.deep.equal([ 'http' ]);
+            expect(client.defaults.auth).to.deep.equal({ username: undefined, password: undefined });
+
+            expect(client.defaults.proxy).to.equal(undefined);
+        });
+    });
+
+    describe('when the artifact registry works over HTTPS', () => {
+
+        const repository = new URL('https://artifactory.example.org');
+
+        const cafile = resolve(__dirname, 'dummy.ca');
+        const ca = readFileSync(cafile).toString('utf8');
+
+        describe('when no proxy is needed', () => {
+
+            it('uses a HTTPS Agent configured to ignored SSL errors', () => {
+                const env = { npm_config_ca: ca };
+                const ignoreSsl = true;
+
+                const client = axiosClient(repository, ignoreSsl, env, repositoryNoAuth);
+
+                expect(client.defaults.baseURL).to.equal(repository.toString());
+                expect(client.defaults.adapter).to.deep.equal([ 'http' ]);
+                expect(client.defaults.auth).to.deep.equal({ username: undefined, password: undefined });
+
+                expect(client.defaults.proxy).to.equal(false);
+
+                expect(client.defaults.httpsAgent.options.ca).to.equal(ca);
+                expect(client.defaults.httpsAgent.options.rejectUnauthorized).to.equal(!ignoreSsl);
+            });
+
+            it('uses a HTTPS Agent configured to not ignored SSL errors', () => {
+                const env = { npm_config_ca: ca };
+                const ignoreSsl = false;
+
+                const client = axiosClient(repository, ignoreSsl, env, repositoryNoAuth);
+
+                expect(client.defaults.baseURL).to.equal(repository.toString());
+                expect(client.defaults.adapter).to.deep.equal([ 'http' ]);
+                expect(client.defaults.auth).to.deep.equal({ username: undefined, password: undefined });
+
+                expect(client.defaults.proxy).to.equal(false);
+
+                expect(client.defaults.httpsAgent.options.ca).to.equal(ca);
+                expect(client.defaults.httpsAgent.options.rejectUnauthorized).to.equal(!ignoreSsl);
+            });
+
+            it('uses a HTTPS Agent configured without a CA', () => {
+                const ignoreSsl = false;
+                const env = {};
+
+                const client = axiosClient(repository, ignoreSsl, env, repositoryNoAuth);
+
+                expect(client.defaults.baseURL).to.equal(repository.toString());
+                expect(client.defaults.adapter).to.deep.equal([ 'http' ]);
+                expect(client.defaults.auth).to.deep.equal({ username: undefined, password: undefined });
+
+                expect(client.defaults.proxy).to.equal(false);
+
+                expect(client.defaults.httpsAgent.options.ca).to.equal(undefined);
+                expect(client.defaults.httpsAgent.options.rejectUnauthorized).to.equal(true);
+            });
+
+            it('uses a HTTPS Agent configured with CA from file', () => {
+                const ignoreSsl = false;
+                const env = {
+                    npm_config_cafile: cafile
+                };
+
+                const client = axiosClient(repository, ignoreSsl, env, repositoryNoAuth);
+
+                expect(client.defaults.baseURL).to.equal(repository.toString());
+                expect(client.defaults.adapter).to.deep.equal([ 'http' ]);
+                expect(client.defaults.auth).to.deep.equal({ username: undefined, password: undefined });
+
+                expect(client.defaults.proxy).to.equal(false);
+
+                expect(client.defaults.httpsAgent.options.ca).to.equal(ca);
+                expect(client.defaults.httpsAgent.options.rejectUnauthorized).to.equal(true);
+            });
+
+            it(`complains if the CA file doesn't exist`, () => {
+                const ignoreSsl = false;
+                const env = {
+                    npm_config_cafile: '/invalid/path/to.ca'
+                };
+
+                expect(() => axiosClient(repository, ignoreSsl, env, repositoryNoAuth))
+                    .to.throw(ConfigurationError, `Could not read npm_config_cafile at /invalid/path/to.ca`);
+            });
+        });
+
+        describe('when a proxy is needed', () => {
+
+            it('uses a HttpsProxyAgent with a HTTPS proxy', () => {
+                const env = {
+                    npm_config_ca: ca,
+                    HTTPS_PROXY: 'https://proxy.example.org'
+                };
+                const ignoreSsl = false;
+
+                const client = axiosClient(repository, ignoreSsl, env, repositoryNoAuth);
+
+                expect(client.defaults.baseURL).to.equal(repository.toString());
+                expect(client.defaults.adapter).to.deep.equal([ 'http' ]);
+                expect(client.defaults.auth).to.deep.equal({ username: undefined, password: undefined });
+
+                expect(client.defaults.proxy).to.equal(false);
+
+                const options = client.defaults.httpsAgent.connectOpts;
+
+                expect(options.ca).to.equal(ca);
+                expect(options.rejectUnauthorized).to.equal(!ignoreSsl);
+                expect(options.host).to.equal('proxy.example.org');
+                expect(options.port).to.equal(443);
+            });
+
+            it('uses a HttpsProxyAgent with a HTTP proxy', () => {
+                const env = {
+                    npm_config_ca: ca,
+                    HTTPS_PROXY: 'http://proxy.example.org'
+                };
+                const ignoreSsl = false;
+
+                const client = axiosClient(repository, ignoreSsl, env, repositoryNoAuth);
+
+                expect(client.defaults.baseURL).to.equal(repository.toString());
+                expect(client.defaults.adapter).to.deep.equal([ 'http' ]);
+                expect(client.defaults.auth).to.deep.equal({ username: undefined, password: undefined });
+
+                expect(client.defaults.proxy).to.equal(false);
+
+                const options = client.defaults.httpsAgent.connectOpts;
+
+                expect(options.ca).to.equal(ca);
+                expect(options.rejectUnauthorized).to.equal(!ignoreSsl);
+                expect(options.host).to.equal('proxy.example.org');
+                expect(options.port).to.equal(80);
+            });
+
+            it('uses a HttpsProxyAgent with a HTTP proxy and a custom port', () => {
+                const env = {
+                    npm_config_ca: ca,
+                    HTTPS_PROXY: 'http://proxy.example.org:8080'
+                };
+                const ignoreSsl = false;
+
+                const client = axiosClient(repository, ignoreSsl, env, repositoryNoAuth);
+
+                expect(client.defaults.baseURL).to.equal(repository.toString());
+                expect(client.defaults.adapter).to.deep.equal([ 'http' ]);
+                expect(client.defaults.auth).to.deep.equal({ username: undefined, password: undefined });
+
+                expect(client.defaults.proxy).to.equal(false);
+
+                const options = client.defaults.httpsAgent.connectOpts;
+
+                expect(options.ca).to.equal(ca);
+                expect(options.rejectUnauthorized).to.equal(!ignoreSsl);
+                expect(options.host).to.equal('proxy.example.org');
+                expect(options.port).to.equal(8080);
+            });
+        });
+    });
+});

--- a/packages/serenity-bdd/spec/cli/io/dummy.ca
+++ b/packages/serenity-bdd/spec/cli/io/dummy.ca
@@ -1,0 +1,4 @@
+-----BEGIN CERTIFICATE-----
+XXXX
+XXXX
+-----END CERTIFICATE-----

--- a/packages/serenity-bdd/src/cli/io/axiosClient.ts
+++ b/packages/serenity-bdd/src/cli/io/axiosClient.ts
@@ -3,11 +3,10 @@ import type { AxiosInstance, AxiosProxyConfig, AxiosRequestConfig } from 'axios'
 import axios from 'axios';
 import * as fs from 'fs';
 import * as https from 'https';
+import { HttpsProxyAgent } from 'https-proxy-agent';
 import { URL } from 'url';
 
 import type { Credentials } from '../model';
-
-const HttpsProxyAgent = require('https-proxy-agent');   // eslint-disable-line @typescript-eslint/no-var-requires
 
 /**
  * @package
@@ -32,13 +31,13 @@ export function axiosClient(
     const options: AxiosRequestConfig = {
         baseURL:    repository.toString(),
         adapter:    [ 'http' ],
-        auth:       repositoryAuth,
+        auth:       repositoryAuth.toJSON() as { username: string, password: string },
     };
 
     if (isHttps(repository.protocol)) {
         options.proxy = false;
         options.httpsAgent = shouldProxy(repository, configuredProxy, configuredNoProxy)
-            ? proxiedHttpsAgent(proxyConfigFrom(configuredProxy), configuredCa, rejectUnauthorized)
+            ? proxiedHttpsAgent(configuredProxy.value, configuredCa, rejectUnauthorized)
             : httpsAgent(configuredCa, rejectUnauthorized);
     }
     else {
@@ -84,10 +83,6 @@ export function shouldProxy(url: URL, configuredProxy: EnvVar, configuredNoProxy
 
 function proxyConfigFrom(proxyUrl: EnvVar): AxiosProxyConfig | undefined {
 
-    if (! proxyUrl || ! proxyUrl.value) {
-        return undefined;
-    }
-
     let parsed: URL;
 
     try {
@@ -120,8 +115,8 @@ function shouldRejectUnauthorizedCertificates(env: { [key: string]: string }, ig
         || env.npm_config_strict_ssl === 'true'
 }
 
-function proxiedHttpsAgent(proxyConfig: AxiosProxyConfig, ca: string, rejectUnauthorized: boolean) {
-    return new HttpsProxyAgent({ ca, rejectUnauthorized, ...proxyConfig });
+function proxiedHttpsAgent(configuredProxy: string, ca: string, rejectUnauthorized: boolean) {
+    return new HttpsProxyAgent(configuredProxy, { ca, rejectUnauthorized });
 }
 
 function httpsAgent(ca: string, rejectUnauthorized: boolean): https.Agent {

